File: day-11/part-1.py
import sys
serial = int(sys.argv[1])

def compute(x, y):
  id = x + 10
  a = (id * y + serial) * id
  b = (a / 100) % 10
  return b - 5

offsets = [-1, 0, 1]
best = -1;
coords = None
for x in range(298):
  for y in range(298):
    x += 2
    y += 2
    val = 0
    for offx in offsets:
      for offy in offsets:
        val += compute(x + offx, y + offy)

    if val > best:
      best = val
      coords = (x - 1, y - 1)

print coords

File: day-11/Part1.java
import java.util.stream.IntStream;

/**
 *
 */
public final class Part1 {

  private Part1() {
  }

  static int serial = 7403;

  public static int compute(int x, int y) {
    int id = x + 10;
    int a = (id * y + serial) * id;
    int b = (a / 100) % 10;
    return b - 5;
  }

  public static int computeLayer(int x, int y, int size) {
    int res = 0;
    res += IntStream.range(0, size).map(i -> compute(x + i, y + size - 1)).sum();
    res += IntStream.range(0, size - 1).map(i -> compute(x + size - 1, y + i)).sum();
    return res;
  }

  static class Result {
    public int x, y, size, score = -1;
  }

  public static void main(String[] args) {
    
    //for (int x = 1; x <= 298; x++) {
    Result res =
    IntStream.range(1, 299).parallel().mapToObj(x -> {
      Result result = new Result();
      for (int y = 1; y <= 298; y++) {
        int tmp = compute(x, y) + computeLayer(x, y, 2);
        for (int i = 3; i <= 301; i++) {
          if (x + i > 300) {
            break;
          }
          tmp += computeLayer(x, y, i);
          if (tmp > result.score) {
            result.score = tmp;
            result.x = x;
            result.y = y;
            result.size = i;
          }
        }
      }
      return result;
    }).reduce((a, b) -> a.score > b.score ? a : b).get();

    System.out.println(res.x + "," + res.y + "," + res.size);
  }
}

